### PR TITLE
RavenDB-15784: Rename attachment of conflicted document if it has dup…

### PR DESCRIPTION
**Pull request by garayx**
…licate name

### CODE REVIEW

File: src/Raven.Server/Documents/DocumentPutAction.cs [hunk 1]
```diff
@@ -422,6 +424,33 @@ private static void ThrowInvalidDocumentId(string id)
...
+                    if (metadata.TryGet(Constants.Documents.Metadata.Attachments, out BlittableJsonReaderArray currentDocumentAttachments))
+                    {
+                        // actualAttachments can contain attachments of current document which were directly put, we want to check if there are attachment with duplicate name
+                        var attachmentsChanged = false;
+                        foreach (BlittableJsonReaderObject bjro in currentDocumentAttachments)
+                        {
+                            var currentAttachment = JsonDeserializationClient.AttachmentName(bjro);
+
+                            if (actualAttachments.Items.Count(o =>
+                                o is DynamicJsonValue attachment && attachment[nameof(AttachmentName.Name)] is LazyStringValue lzv && lzv == currentAttachment.Name) > 1)
```
Comment (ayende) [inline src/Raven.Server/Documents/DocumentPutAction.cs]: The types here are _very_ exact. Can it be other things? 
This looks very fragile.

File: src/Raven.Server/Documents/AttachmentsStorage.cs [hunk 2]
```diff
@@ -901,6 +901,64 @@ public AttachmentDetails MoveAttachment(DocumentsOperationContext context, strin
             return result;
         }
 
+        public AttachmentDetails RenameAttachment(DocumentsOperationContext context, Slice lowerId, string name, string hash, string contentType, AttachmentType attachmentType, string newName)
```
Comment (ayende) [inline src/Raven.Server/Documents/AttachmentsStorage.cs:925]: Why is this required? The tag on the attachment is for debug only.
We don't actually need to rename it.

File: src/Raven.Server/Documents/DocumentPutAction.cs [hunk 3]
```diff
@@ -422,6 +424,33 @@ private static void ThrowInvalidDocumentId(string id)
...
+                            {
+                                // rename currentAttachment which is already saved in storage
+                                _documentsStorage.AttachmentsStorage.RenameAttachment(
```
Comment (ayende) [inline src/Raven.Server/Documents/DocumentPutAction.cs]: What if we have three duplicates?

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Documents/DocumentPutAction.cs (hunk 1 above)
Comment (garayx) [reply src/Raven.Server/Documents/DocumentPutAction.cs]: this is returned from the attachments storage:
https://github.com/ravendb/ravendb/blob/ff622ad5b817a4245e8011e0a5779599236f1511/src/Raven.Server/Documents/AttachmentsStorage.cs#L146
and the values are from table

File: src/Raven.Server/Documents/AttachmentsStorage.cs (hunk 2 above)
Comment (garayx) [reply src/Raven.Server/Documents/AttachmentsStorage.cs:925]: yes the unique value is key, but we don't allow attachments to have same name when it is added from studio or Client API. Conflicts in replication can cause such scenario, which in place can break other things like patch.

Comment (ayende) [reply src/Raven.Server/Documents/AttachmentsStorage.cs:925]: What happens if we have _another_ conflict? Then we'll have a duplicate on `resolved_name`, no?

File: src/Raven.Server/Documents/DocumentPutAction.cs (hunk 1 above)
Comment (ayende) [reply src/Raven.Server/Documents/DocumentPutAction.cs]: That is fragile. If we'll change this in the future, there is no indication that this is related.

File: src/Raven.Server/Documents/DocumentPutAction.cs (hunk 3 above)
Comment (garayx) [reply src/Raven.Server/Documents/DocumentPutAction.cs]: The only chance for this is when the doc already has attachments with duplicate names (got conflicted before this PR merged) and gets into conflict again after this PR merged. For normal doc, after this PR there is always going to be max 2 duplicates.

We can treat this issue by handling here the > 2 duplicates case or do schema update to rename attachments with duplicate  name

File: src/Raven.Server/Documents/DocumentPutAction.cs (hunk 1 above)
Comment (garayx) [reply src/Raven.Server/Documents/DocumentPutAction.cs]: done

File: src/Raven.Server/Documents/AttachmentsStorage.cs (hunk 2 above)
Comment (garayx) [reply src/Raven.Server/Documents/AttachmentsStorage.cs:925]: done

**Review by ayende — APPROVED** [review]

File: src/Raven.Server/Documents/Replication/ConflictManager.cs [hunk 4]
```diff
@@ -85,9 +86,16 @@ public ConflictManager(DocumentDatabase database, ResolveConflictOnReplicationCo
...
-
-                    return;
+                    try
+                    {
+                        _conflictResolver.PutResolvedDocument(documentsContext, resolved, conflictedDoc, localDocumentTuple.Document);
+                        return;
+                    }
+                    catch (AttachmentConflictException e)
+                    {
+                        if (_log.IsInfoEnabled)
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/ConflictManager.cs:96]: Under what scenario do we want this behavior? We should always resolve things automatically

File: src/Raven.Server/Documents/Replication/ConflictManager.cs [hunk 5]
```diff
@@ -137,14 +146,26 @@ public static void AssertChangeVectorNotNull(string conflictedChangeVector)
                 InvalidConflictWhenThereIsNone(conflict.Id);
 
             conflictedDocs.Add(conflict.Clone());
+            conflictedDocs = conflictedDocs.OrderBy(x => x.Doc?.GetHashCode()).ToList();
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/ConflictManager.cs:149]: This is effectively random, why do you do this?
Also, it isn't guaranteed to be the same on different nodes

File: src/Raven.Server/Documents/Replication/ConflictManager.cs [hunk 6]
```diff
@@ -137,14 +146,26 @@ public static void AssertChangeVectorNotNull(string conflictedChangeVector)
...
                 scriptResolver,
                 conflictedDocs,
                 documentsContext.GetLazyString(collection), out var resolved))
             {
-                _conflictResolver.PutResolvedDocument(documentsContext, resolved, conflict);
+                try
+                {
+                    _conflictResolver.PutResolvedDocument(documentsContext, resolved, conflict, relevantLocalDoc.Document);
+                }
+                catch (AttachmentConflictException e)
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/ConflictManager.cs:161]: I don't think we want this

File: src/Raven.Server/Documents/Replication/ResolveConflictOnReplicationConfigurationChange.cs [hunk 7]
```diff
@@ -400,6 +412,125 @@ public DocumentConflict ResolveToLatest(List<DocumentConflict> conflicts)
...
+                        }
+
+                        foreach (var match in matches)
+                        {
+                            // find which name of the matched attachments doesn't exist in the current metadata
+                            if (CountInAttachmentsList(match.Name, currentDocumentAttachments) == 0)
+                            {
+                                if (_database.DocumentsStorage.AttachmentsStorage.RenameAttachment(context, lowerId, match.Name, match.Hash, match.ContentType, AttachmentType.Document, currentAttachment.Name) == null)
+                                    throw new AttachmentConflictException($"Could not rename attachment '{match.Name}' to '{currentAttachment.Name}' in attachments storage of document '{lowerId}', attachment has the following: Hash: '{currentAttachment.Hash}', ContentType: '{currentAttachment.ContentType}'.", lowerId.ToString());
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/ResolveConflictOnReplicationConfigurationChange.cs:453]: Can use `_1`, `_2`, etc instead.
Can use random value, etc.

We can't throw here.

File: src/Raven.Server/Documents/Replication/ResolveConflictOnReplicationConfigurationChange.cs [hunk 8]
```diff
@@ -400,6 +412,125 @@ public DocumentConflict ResolveToLatest(List<DocumentConflict> conflicts)
...
+                        RemoveFromAttachmentsToRemoveListIfNeeded(currentAttachment.Name);
+                        continue;
+                    case 0:
+                        // attachment exists in metadata, does not exist in storage (renamed/added in conflict resolution)
+                        // try to find attachments in attachment storage with both same hash and content type
+                        var matches = storageAttachmentsDetails.Where(x => x.Hash == currentAttachment.Hash && x.ContentType == currentAttachment.ContentType).ToList();
+                        if (matches.Count == 0)
+                        {
+                            // attachment is in the metadata but doesn't exist in the attachments storage
+                            throw new AttachmentConflictException($"Attachment '{currentAttachment.Name}' was found in '{nameof(DocumentFlags.Resolved)}' document '{lowerId}' metadata, but was not found in the attachments storage using the following: Hash: '{currentAttachment.Hash}', ContentType: '{currentAttachment.ContentType}'.", lowerId.ToString());
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/ResolveConflictOnReplicationConfigurationChange.cs:444]: Shouldn't this be a missing attachment error?

File: src/Raven.Server/Documents/Replication/ResolveConflictOnReplicationConfigurationChange.cs [hunk 9]
```diff
@@ -400,6 +412,125 @@ public DocumentConflict ResolveToLatest(List<DocumentConflict> conflicts)
...
+                                
+                                RemoveFromAttachmentsToRemoveListIfNeeded(match.Name);
+                                break;
+                            }
+                        }
+
+                        break;
+                    case 2:
+                        // there are attachments with duplicate names in storage
+                        var duplicates = storageAttachmentsDetails.Where(x => x.Name == currentAttachment.Name && (x.Hash != currentAttachment.Hash || x.ContentType != currentAttachment.ContentType)).ToList();
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/ResolveConflictOnReplicationConfigurationChange.cs:463]: Why do you care about the attachment name? Only the hash matters to us

**Review by ayende — CHANGES_REQUESTED** [review]
Talk to me about this, please

**Comment by ppekrol** [comment]
This will be handled differently. @karmeli87 and @garayx spoke.

File: src/Raven.Server/Documents/Replication/ConflictManager.cs (hunk 5 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/Replication/ConflictManager.cs:149]: Right... we need some deterministic way to order them.
Maybe rely on the CV for that? (But I guess we need to order it internally as well)

Comment (ayende) [reply src/Raven.Server/Documents/Replication/ConflictManager.cs:149]: Ordering by CV will be a good idea, yes.

File: src/Raven.Server/Documents/Replication/ResolveConflictOnReplicationConfigurationChange.cs (hunk 8 above)
Comment (garayx) [reply src/Raven.Server/Documents/Replication/ResolveConflictOnReplicationConfigurationChange.cs:444]: This is to prevent from adding an attachment in conflict resolver script
https://github.com/ravendb/ravendb/pull/11332/files#diff-0274c68798c060bc7f04679d2b76c93d9d1f92404872601fcf7d4e7f547b224fR1951

File: src/Raven.Server/Documents/Replication/ResolveConflictOnReplicationConfigurationChange.cs (hunk 9 above)
Comment (garayx) [reply src/Raven.Server/Documents/Replication/ResolveConflictOnReplicationConfigurationChange.cs:463]: The duplicate names was the original issue we had to solve, because duplicate names can break patch and should be disallowed

File: src/Raven.Server/Documents/Replication/ResolveConflictOnReplicationConfigurationChange.cs (hunk 7 above)
Comment (garayx) [reply src/Raven.Server/Documents/Replication/ResolveConflictOnReplicationConfigurationChange.cs:453]: `RenameAttachment` should not fail, its a precaution so it can be resolved manually if something goes wrong and we cannot add 1,2, random value, because this is result of script of the `ScriptResolver`

File: src/Raven.Server/Documents/Replication/ConflictManager.cs (hunk 6 above)
Comment (garayx) [reply src/Raven.Server/Documents/Replication/ConflictManager.cs:161]: this is to mark the doc as conflicted

**Comment by garayx** [comment]
https://github.com/ravendb/ravendb/pull/11434
